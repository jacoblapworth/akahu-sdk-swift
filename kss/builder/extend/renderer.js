const { regexDictionary, types } = require('../parser/constants');
const tinycolor = require('tinycolor2');

module.exports = function(handlebars) {
	/**
	 * Extracts the variable properties from the line
	 * returns:
	 *  token name, token value, and optional [ token example ]
	 */
	handlebars.registerHelper('renderTokens', renderTokens);

	/**
	 * Compares the name to see if the container needs to be rendered as inverted.
	 * Similar to the isEqual
	 *
	 * e.g.
	 * {{#isInverted name}}
	 * 		ADD THIS
	 * {{else}} [optional]
	 * 		WILL EXECUTE THIS
	 * {{/isInverted}}
	 */
	handlebars.registerHelper('isInverted', isInverted);
};

const testString = 'To understand what recursion is, you must first understand recursion.';


function renderTokens(variables, block) {
	var newBlock = '';
	variables.split('\n').forEach(line => {
		var test;
		const matches = line.match(/[^:]+/g).map(match => match.trim());

		if (matches[1]) {
			this.token = {};
			this.token.name = matches[0];
			this.token.value = matches[1];
			this.token.title = matches[3];
			this.token.description = matches[4];
			if (matches[2] != undefined) {
				this.token.example = getExample(matches[2], this.token);
			}
			newBlock += block.fn(this);
		}
	});
	return newBlock;
}

function getExample(type, token) {
	switch(type) {
		case types.color:
			const codeTextClass = tinycolor(token.value).isDark() ?
				'ds-colour-code-light'
				: null;
			return `
					<div class="ds-colour-example" style="background-color:${token.value}">
						<code class="${codeTextClass}">${token.value}</code>
						<code class="${codeTextClass}">${token.name}</code>
					</div>`;
		case types.border:
			return `
				<div class="ds-colour-example" style="border:${token.value}">
					<code>${token.name}</code>
				</div>`;
		case types.shadow:
			return `
					<div class="ds-colour-example" style="box-shadow:${token.value}">
						<code>${token.name}</code>
					</div>`;
		case types.fontSize:
			return  `<span class="xui-text-truncated" style="font-size:${token.value};">Aa</span>`;
		case types.fontWeight:
			return `<span style="font-weight:${token.value};">Aa</span>`;
		case types.lineHeight:
			return `<span style="line-height:${token.value}">${testString}</span>`;
		case types.spacing:
			return `<div class="ds-spacing-example" style="height:${token.value};width:${token.value};"></div>`;
		default:
			return '';
	}
}

function isInverted(name, options) {
	return regexDictionary.inverted.test(name) ? options.fn(this) : options.inverse(this);
}
