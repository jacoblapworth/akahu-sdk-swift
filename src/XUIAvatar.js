import React from 'react';
import cn from 'classnames';
import XUIClasses from 'xui-css-classes';
import 'String.prototype.at';

/**
 * @public
 *
 * Property types for this component
 */
const propTypes = {

	className: React.PropTypes.string,
	qaHook: React.PropTypes.string,

	/** @property {String} [value] The text to display in the avatar */
	value: function(props, propName) {
		if (!props[propName] && !props.imageUrl) {
			return new Error('Avatar component requires either a non-empty `value` or `imageUrl` property');
		}
	},

	/** @property {String} [imageUrl] the image the component should render. Initials rendered otherwise */
	imageUrl: React.PropTypes.string,

	/** @property {String} [size=medium] The size of the avatar. Can be small, medium or large */
	size: React.PropTypes.oneOf(['small', 'medium', 'large']),

	/** @property {String} [identifier] A unique string that will be used to generate the color of the avatar if color is not provided. If this is not set then value is used as the identifier. */
	identifier: React.PropTypes.string,

	/** @property {Function} [onError] Error handler if the avatar image fails to load */
	onError: React.PropTypes.fn
};

/**
 * @public
 * Default property values for this component
 */
const defaultProps = {
	size: 'medium',
	pureRender: true
};

/**
 * @private
 * Map of sizes to XUI Classes properties
 */
const sizeMap = {
	'small': 'SMALL',
	'medium': '',
	'large': 'LARGE'
};

/**
 * @private
 * Gets the avatar color class based on the identifier
 * @param {String} identifier
 * @returns {String} The class name to use for the avatar's color
 */
function getAvatarColorClass(identifier) {
	let i;
	let c;
	let h = 5381;
	const len = identifier.length >>> 0;

	for (i = 0; i < len; i++) {
		c = identifier.charCodeAt(i);
		h = (((h << 5) + h) ^ c) >>> 0;
	}

	// The XUI class names are not zero-indexed
	const colorNumber = (h % XUIClasses.Avatar.COLORS.length);
	return XUIClasses.Avatar.COLORS[colorNumber];
}

export default function Avatar(props) {
	const { qaHook, imageUrl, size, identifier, value, onError } = props;

	const avatarClassNames = cn(
		props.className,
		XUIClasses.Avatar.BASE,
		XUIClasses.Avatar[sizeMap[size]],
		imageUrl ? null : getAvatarColorClass(identifier || value)
	);

	if (imageUrl) {
		return <img onError={onError} data-automationid={qaHook} className={avatarClassNames} role="presentation" alt="" src={imageUrl}/>;
	} else {
		return <abbr data-automationid={qaHook} className={avatarClassNames} role="presentation">{props.value.at(0)}</abbr>;
	}
}

Avatar.propTypes = propTypes;
Avatar.defaultProps = defaultProps;
