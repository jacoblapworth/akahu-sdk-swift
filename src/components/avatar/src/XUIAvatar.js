import React from 'react';
import XUISimpleAvatar from './XUISimpleAvatar';

const propTypes = {
	/** @property {Function} [onError] Function to be used if the avatar renders an image and the image load fails  */
	onError: React.PropTypes.func,

	/** @property {String} value The text to display in the avatar. In the simple avatar component, it is optional, but here it is mandatory */
	value: React.PropTypes.string.isRequired
};

/**
 * @private
 * onError handler for the image element
 * @param {Object} e Error object
 */
function onError(e) {
	const { props } = this;

	this.setState({
		imageError: true
	});

	if(props.onError) {
		props.onError(e);
	}
}

export default class XUIAvatar extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			imageError: false
		};
	}
	render() {
		const { props, state } = this;

		if (state.imageError) {
			return <XUISimpleAvatar {...props} imageUrl={null} />
		} else {
			const errorHandler = onError.bind(this);
			return <XUISimpleAvatar {...props} onError={errorHandler} />
		}
	}
}

XUIAvatar.propTypes = propTypes;
