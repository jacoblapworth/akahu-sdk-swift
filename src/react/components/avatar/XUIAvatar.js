import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import memoizeOne from 'memoize-one';
import cn from 'classnames';
import '../helpers/xuiGlobalChecks';
import { sizeClassNames, classNames, variantClassNames } from './constants';
import { getAvatarColorClass, abbreviateAvatar } from './utils';

export default class XUIAvatar extends PureComponent {
  state = {
    imageLoaded: false,
  };

  /**
   * onError handler for the image element
   * @param {Error} e Error object
   */
  onError = e => {
    const { onError } = this.props;
    onError && onError(e);
  };

  onLoad = () => {
    this.setState({
      imageLoaded: true,
    });
  };

  memoizedAbbreviateAvatar = memoizeOne(abbreviateAvatar);

  render() {
    const { imageLoaded } = this.state;

    const { qaHook, className, imageUrl, size, identifier, value, variant } = this.props;

    const avatarClassNames = cn(
      className,
      classNames.base,
      sizeClassNames[size],
      variantClassNames[variant],
    );

    const avatarCharacterCount = variant === 'business' && size !== '2xsmall' ? 3 : 2; // 2xsmall cannot fit 3 characters without overflowing
    return (
      <React.Fragment>
        {imageUrl && (
          <img
            alt=""
            className={cn(avatarClassNames, { [classNames.hidden]: !imageLoaded })}
            data-automationid={qaHook}
            onError={this.onError}
            onLoad={this.onLoad}
            src={imageUrl}
          />
        )}
        <abbr
          className={cn(avatarClassNames, getAvatarColorClass(identifier || value), {
            [classNames.hidden]: imageUrl && imageLoaded,
          })}
          data-automationid={qaHook}
          role="presentation"
        >
          {this.memoizedAbbreviateAvatar(value, avatarCharacterCount)}
        </abbr>
      </React.Fragment>
    );
  }
}

XUIAvatar.propTypes = {
  className: PropTypes.string,
  qaHook: PropTypes.string,

  /** The avatar variant */
  variant: PropTypes.oneOf(Object.keys(variantClassNames)),

  /** The text to display in the avatar */
  value: PropTypes.string.isRequired,

  /** the image the component should render. Initials rendered otherwise */
  imageUrl: PropTypes.string,

  /** The size of the avatar. Can be small, medium, large or xlarge */
  size: PropTypes.oneOf(Object.keys(sizeClassNames)),

  /** A unique string that will be used to generate the color of the avatar if color is not
   * provided. If this is not set then value is used as the identifier. */
  identifier: PropTypes.string,

  /** Error handler if the avatar image fails to load */
  onError: PropTypes.func,
};

XUIAvatar.defaultProps = {
  size: 'medium',
};
