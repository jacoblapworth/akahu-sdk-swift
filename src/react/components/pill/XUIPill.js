import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import crossSmall from '@xero/xui-icon/icons/cross-small';

import XUIIcon from '../icon/XUIIcon';
import XUIButton from '../button/XUIButton';
import InnerPill from './InnerPill';

/**
 * @private
 * @param {Function} callback
 * @param {Object} scope
 * @return {Function} the provided callback with the scope passed in as an argument.
 */
function returnCallbackWithScope(callback, scope) {
	return () => callback(scope);
}

export default class XUIPill extends Component {

	constructor(props) {
		super(props);

		this.state = {
			isFocussed: false
		};

		this.toggleFocus = this.toggleFocus.bind(this);
	}

	toggleFocus() {
		this.setState(prevState => ({
			isFocussed: !prevState.isFocussed
		}));
	}

	render() {
		const pill = this;

		const {
			avatarProps,
			className,
			deleteButtonLabel,
			hasLayout,
			href,
			isInvalid,
			onClick,
			onDeleteClick,
			qaHook,
			secondaryText,
			target,
			value
		} = pill.props;

		const baseClass = 'xui-newpill';

		const pillClasses = cn(
			className,
			baseClass,
			{
				[`${baseClass}-layout`]: hasLayout,
				[`${baseClass}-is-invalid`] : isInvalid,
				[`${baseClass}-is-focussed`]: pill.state.isFocussed,
				[`${baseClass}-is-deleteable`]: onDeleteClick
			}
		);

		const innerPillProps = {
			avatarProps,
			href,
			onClick,
			qaHook,
			secondaryText,
			target,
			value
		};

		const onDeleteCallback = onDeleteClick && returnCallbackWithScope(onDeleteClick, pill);

		const deleteButton = onDeleteCallback && (
			<XUIButton
				className={`${baseClass}--button-icon`}
				variant="icon"
				onClick={onDeleteCallback}
				title={deleteButtonLabel}
				aria-label={deleteButtonLabel}
				qaHook={qaHook ? `deletePillButton-${qaHook}` : null}
			>
				<XUIIcon path={crossSmall} />
			</XUIButton>
		);

		return (
			<div
				className={pillClasses}
				onFocus={pill.toggleFocus}
				onBlur={pill.toggleFocus}
				data-automationid={qaHook}>
				<InnerPill {...innerPillProps} />
				{deleteButton}
			</div>
		)
	}
}

const noop = () => {};

XUIPill.defaultProps = {
	deleteButtonLabel: 'Delete',
	hasLayout: true,
	onClick: noop
};

XUIPill.propTypes = {
	/** @property {Object} [avatarProps] Props for the avatar to be displayed, must adhere to the XUIAvatar component API described at https://github.dev.xero.com/UXE/xui-avatar. Version 6.0.0+. Not providing props will omit the avatar entirely. */
	avatarProps: PropTypes.object,
	/** @property {String} [className] Apply classes to the outer Pill `div` element. */
	className: PropTypes.string,
	/** @property {String} [deleteButtonLabel='Delete'] Specify an alternate label attribute for the delete button, defaults to 'Delete'. */
	deleteButtonLabel: PropTypes.string,
	/** @property {Boolean} [hasLayout=true] Remove the XUI layout class by specifiying false. */
	hasLayout: PropTypes.bool,
	/** @property {String} [href] This will make the value an `anchor` element instead of a `span` element and adds the href as the link. */
	href: PropTypes.string,
	/** @property {String} [isInvalid] When invalid, displays the text in a red colour. */
	isInvalid: PropTypes.bool,
	/** @property {Function} [onClick=noop] Callback to fire when the main pill content is clicked. */
	onClick: PropTypes.func,
	/** @property {Function} [onDeleteClick] Callback to fire when the delete pill button is clicked. When omitted, the delete button is also ommitted from the view. */
	onDeleteClick: PropTypes.func,
	/** @property {String} [qaHook] Adds a data attribute hook for testing. The inner pill has the format `pillButton-${YOUR_STRING_GOES_HERE}`. The delete button has the format `deletePillButton-${YOUR_STRING_GOES_HERE}`. The wrapper pill div has the format `${YOUR_STRING_GOES_HERE}`. */
	qaHook: PropTypes.string,
	/** @property {Boolean} [target] When an `href` is supplied, adds a target attribute, else is ignored. */
	target: PropTypes.string,
	/** @property {String} [secondaryText] Adds a muted secondary text for the pill, appears before the main value. */
	secondaryText: PropTypes.string,
	/** @property {String} [value] The text to display inside the pill. */
	value: PropTypes.string
};
