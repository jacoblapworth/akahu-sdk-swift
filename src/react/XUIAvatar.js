import XUIAvatarGroup from './components/avatar/XUIAvatarGroup';
import XUISimpleAvatar from './components/avatar/XUISimpleAvatar';
import XUIAvatarCounter from './components/avatar/XUIAvatarCounter';
import XUIAvatar from './components/avatar/XUIAvatar';

export { XUIAvatar as default, XUIAvatarGroup, XUIAvatarCounter, XUISimpleAvatar };
