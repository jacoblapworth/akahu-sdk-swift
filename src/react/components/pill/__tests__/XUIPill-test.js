import React from 'react';
import Enzyme, { shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import XUIPill from '../XUIPill';
import XUIAvatar from '../../avatar/XUIAvatar';

Enzyme.configure({ adapter: new Adapter() });

const NOOP = () => {};

describe('<XUIPill />', () => {

	it('renders the pill with the specified value prop', () => {
		const pill = renderer.create(
			<XUIPill value="Value Pill" />
		);

		expect(pill).toMatchSnapshot();
	});

	it('renders the pill with the specified className prop', () => {
		const pill = render(
			<XUIPill className='xui-test-class' />
		);

		expect(pill.hasClass('xui-test-class')).toBeTruthy();
	});

	it('will render the pill text in an "a" tag when a href prop is provided', () => {
		const pill = shallow(
			<XUIPill href="http://xero.com"/>
		);

		expect(pill.find('a')).toBeTruthy();
	});

	it('renders the pill with the invalid class when the isInvalid prop is true', () => {
		const pill = render (
			<XUIPill isInvalid={true} />
		);

		expect(pill.hasClass('xui-pill-is-invalid')).toBeTruthy();
	});

	it('will render avatars when passed as an avatar prop', () => {
		const imageUrl = "https://xui.xero.com/static/xpert-avatar.png";
		const avatarProps = {
			imageUrl,
			size: 'small',
			role: 'presentation',
			value: 'A'
		};
		const pill = shallow(
			<XUIPill avatarProps={avatarProps} />
		);

		expect(pill.find('.xui-avatar')).toBeTruthy();
	});

	it('renders the pill with the provided avatar component', () => {
		const avatar = <XUIAvatar value="value pill" size="small" />
		const pill = renderer.create(
			<XUIPill value="Value Pill" avatar={avatar} />
		);

		expect(pill).toMatchSnapshot();
	});

	it('expect an error if both avatar and avatarProps are provided', () => {
		const imageUrl = "https://xui.xero.com/static/xpert-avatar.png";
		const avatar = <XUIAvatar value="value pill" size="small" />
		const avatarProps = {
			imageUrl,
			size: 'small',
			role: 'presentation',
			value: 'A'
		};
		expect(() => renderer.create(
			<XUIPill value="Value Pill" avatar={avatar} avatarProps={avatarProps} />
		)).toThrow();
	});

	it('invokes the callback passed into the onDeleteClick prop with itself passed in as an argument', () => {
		const callback = jest.fn();
		const pill = mount(
			<XUIPill value="Pill" onDeleteClick={callback} />
		);

		pill.find('.xui-pill--button-icon').first().simulate('click');
		expect(callback.mock.calls.length).toEqual(1);
	});

	it('invokes the callback passed into the onClick prop', () => {
		const callback = jest.fn();
		const pill = mount(
			<XUIPill value="Pill" onClick={callback} />
		);

		pill.find('.xui-button').simulate('click');
		expect(callback.mock.calls.length).toEqual(1);
	});

	it('renders the text passed in the secondaryText prop', () => {
		const pill = mount(
			<XUIPill secondaryText='supplementary' />
		);

		expect(pill.find('.xui-pill--secondary').text()).toEqual('supplementary');
	});

	it('renders itself as focused appropriately', () => {
		const pill = mount(
			<XUIPill />
		);

		pill.childAt(0).simulate('focus');

		expect(pill.childAt(0).hasClass('xui-pill-is-focused')).toBeTruthy();
	});

	it('should render an automation id when a qaHook is passed in', () => {
		const automationid = renderer.create(
			<XUIPill qaHook="pill-test" />
		);

		expect(automationid).toMatchSnapshot();
	});

	it('should render a title when passed', () => {
		const pill = renderer.create(
			<XUIPill title="pill title" />
		);

		expect(pill).toMatchSnapshot();
	});

	it('should render a target when passed in', () => {
		const pill = renderer.create(
			<XUIPill href="http://xero.com" target="_blank" />
		);

		expect(pill).toMatchSnapshot();
	});

	it('should render an error icon when invalid', () => {
		const pill = renderer.create(
			<XUIPill value="Error pill" isInvalid />
		);

		expect(pill).toMatchSnapshot();
	});

	it('should render an error icon when invalid even with an avatar', () => {

		const imageUrl = "https://xui.xero.com/static/xpert-avatar.png";
		const avatarProps = {
			imageUrl,
			size: 'small',
			role: 'presentation',
			value: 'A'
		};

		const pill = renderer.create(
			<XUIPill value="Error pill" isInvalid {...avatarProps} />
		);

		expect(pill).toMatchSnapshot();
	});

	it('should render a label for the delete button when passed in', () => {
		const pill = mount(<XUIPill deleteButtonLabel="alternate delete label" onDeleteClick={NOOP}/>);

		expect(pill.find('.xui-pill--button-icon').first().html()).toContain('title="alternate delete label"');
	});

	it('should render a delete button label of \'Delete\' by default', () => {
		const pill = mount(<XUIPill onDeleteClick={NOOP}/>);

		expect(pill.find('.xui-pill--button-icon').first().html()).toContain('title="Delete"');
	});

	it('should switch the focus state when toggleFocus is called', () => {
		const pill = shallow(<XUIPill />);

		expect(pill.state('isFocused')).toBeFalsy();

		pill.instance().toggleFocus();

		expect(pill.state('isFocused')).toBeTruthy();
	});

	it('should render a pills with correct size modifiers', () => {
		const sizes = ['medium', 'small', 'xsmall'];
		sizes.forEach(size => {
			const pill = renderer.create(
				<XUIPill
					size={size}
					value="Test"
					avatarProps={{
						value: 'Test Render'
					}}
					onDeleteClick={NOOP}
				/>
			);
			expect(pill).toMatchSnapshot();
		});
	});

	it('should render inside a tooltip when using the debug flag', () => {
		const pill = renderer.create(<XUIPill debugShowToolTip/>);

		expect(pill).toMatchSnapshot();
	});

	it('should render a pill with the maxwidth limiting class', () => {
		const pill = renderer.create(<XUIPill isLimitedWidth/>);

		expect(pill).toMatchSnapshot();
	});
})
