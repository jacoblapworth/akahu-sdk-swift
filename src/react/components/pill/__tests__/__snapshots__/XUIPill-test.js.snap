// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUIPill /> renders the pill with the specified value prop 1`] = `
<div
  className="xui-pill xui-pill-layout"
  data-automationid={undefined}
  onBlur={[Function]}
  onFocus={[Function]}
>
  <span
    className="xui-pill--content xui-pill--button"
    data-automationid={undefined}
  >
    <span
      className="xui-pill--text"
    >
      Value Pill
    </span>
  </span>
</div>
`;

exports[`<XUIPill /> should render a target when passed in 1`] = `
<div
  className="xui-pill xui-pill-layout"
  data-automationid={undefined}
  onBlur={[Function]}
  onFocus={[Function]}
>
  <a
    className="xui-button xui-pill--content"
    data-automationid={undefined}
    disabled={false}
    href="http://xero.com"
    onClick={[Function]}
    onKeyDown={undefined}
    onKeyPress={[Function]}
    rel={undefined}
    tabIndex={0}
    target="_blank"
    title={undefined}
  />
</div>
`;

exports[`<XUIPill /> should render a title when passed 1`] = `
<div
  className="xui-pill xui-pill-layout"
  data-automationid={undefined}
  onBlur={[Function]}
  onFocus={[Function]}
>
  <span
    className="xui-pill--content xui-pill--button"
    data-automationid={undefined}
  />
</div>
`;

exports[`<XUIPill /> should render an automation id when a qaHook is passed in 1`] = `
<div
  className="xui-pill xui-pill-layout"
  data-automationid="pill-test"
  onBlur={[Function]}
  onFocus={[Function]}
>
  <span
    className="xui-pill--content xui-pill--button"
    data-automationid="pill-test--inner"
  />
</div>
`;
